Popover: Prevent the Popover from going outside the browser window in some situations

Saman-00 opened the pull request:
## Description
`popover-content` is positioned based on the parent element of the `MudPopover` component, and when this parent element is outside the browser window, the popup also goes outside the window and is partially visible or not visible at all, this seems to be a common occurrence in `MudDataGrid`'s simple filter popover when the grid has a scroll.

This behavior can be viewed here: https://try.mudblazor.com/snippet/cOGeEScJpoYZVWLx
This commit attempts to fix the issue by preventing the popover from going outside the window.
It potentially fixes these issues: Fixes #8180, fixes #8173, fixes #7032, fixes #7707

This video shows how it currently is: https://github.com/MudBlazor/MudBlazor/assets/5804803/803e4884-a5b0-4452-ae4c-afe2552c3ed8
This one shows how it will behave after the change: https://github.com/MudBlazor/MudBlazor/assets/5804803/5f30914b-9bd0-4680-a603-246639298a47


## How Has This Been Tested?
This change has been tested visually, I made a component out of the TryMudBlazor snippet and tested manually.

## Type of Changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->


## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

Saman-00 committed:
Prevent the Popover from going outside the browser window in some situations

danielchalmers commented:
Strangely the notification popover still seems to get clipped: https://github.com/MudBlazor/MudBlazor/issues/7032#issuecomment-2055959684

Shrink the window until it changes layout, click the hamburger menu, then click the notifications icon to see

Saman-00 committed:
fix: popover is positioned wrong when window has scrolled

Saman-00 commented:
> Strangely the notification popover still seems to get clipped ([#7032 (comment)](https://github.com/MudBlazor/MudBlazor/issues/7032#issuecomment-2055959684))
> 
> ![image](https://private-user-images.githubusercontent.com/7112040/325696992-c4a563f3-6cef-43ff-9b62-83c9ad0a3bd0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyODIwMzksIm5iZiI6MTcxNTI4MTczOSwicGF0aCI6Ii83MTEyMDQwLzMyNTY5Njk5Mi1jNGE1NjNmMy02Y2VmLTQzZmYtOWI2Mi04M2M5YWQwYTNiZDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDlUMTkwODU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjI2ODU1YTI0ZDc3YWM0MTFlYTBiNTE3ODU0MDBlMWI4MTQ5MTNhMzk4NGIzZDZkNDFkMGZkOWNiMDZmZjI1ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.VWuHvP1WlHstaBHMwdX7F4Ajo-1CdjXzb7_eq1C4Y6M)
> 
> Shrink the window until it changes layout, click the hamburger menu, then click the notifications icon to see

Do you mean that the popover still goes outside the window after my changes? because I can't reproduce that.

![image](https://github.com/MudBlazor/MudBlazor/assets/5804803/69a7334f-ecc5-4b2f-980f-b3634c454633)

But my notification area looks wrong, maybe that is the reason

danielchalmers commented:
That seems to be working now, not sure if i was on the wrong branch before (🤦 )

![image](https://github.com/MudBlazor/MudBlazor/assets/7112040/7f64f392-93b0-4ac4-992d-f3b0ff146a8b)

Seems to be a miscalculation with the scrolling now:


https://github.com/MudBlazor/MudBlazor/assets/7112040/dbe6341c-a205-4f07-98b7-5f82da533c2a

danielchalmers commented:
What do you think about this by the way https://github.com/MudBlazor/MudBlazor/issues/1167#issuecomment-2075508364? I want to get the tooltip layout fixed but something needs to be adjusted in the javascript (try settings `display: contents` on a tooltip then hover over it).

ScarletKuro commented:
Hi

> Seems to be a miscalculation with the scrolling now:

@Saman-00 can you fix that? Then we can merge if there no other issues.

Saman-00 commented:
That can be fixed, I just need to not add scrolling offset when `position: fixed`.

But now I think my approach to the problem might have been wrong from the beginning.
What this PR tries to achieve is force every popover to be within the viewport, but I don't think that is the intended behavior for popover, if you view the Popover example in the doc with these changes, you will notice that everything looks wrong.

![image](https://github.com/MudBlazor/MudBlazor/assets/5804803/8f879483-da23-43e7-b44a-ff1c282a3122)
All of the popovers jump into view when window resizes.

The only time I can think of where this change would make sense is either when position is fixed, or when `LockScroll="true"`, because in both situations, it doesn't make sense to render the element outside the window, so we force them to stay within the window.
But in case of `LockScroll`, we can't use it for that either because it is an attribute of `Overlay` and not `Popover`, and we can't check the document body's class because the same thing will happen as the image above whenever an Overlay with `LockScroll` parameter is displayed.

Should I just close this PR or add it only when `position: fixed`?
What do you think, do you think there is a better solution?

danielchalmers commented:
@Saman-00 Would it make sense if
- The popover always initially spawns in the viewport
- If you resize the window, it resizes proportionally by dimensions; If it was out of view it stays out, but if it was in view it stays in.
- If you scroll the viewport, it will stay where it is on the page, not the viewport, so you can scroll it out of view

Me and @Garderoben talked about scroll locking before and the current behavior where you can freely scroll was preferred. [MUI does lock it](https://mui.com/material-ui/react-popover/) on the other hand.

ScarletKuro commented:
Maybe we should not even try to fix it, if it's too much work.
There is plan to replace the popover with https://floating-ui.com/
Because nobody wants to maintain our popover js code and I had this arguments to use it:
1) It's used by most websites in the world, so it's always maintained.
2) The JavaScript code is not that large.
3) Our popover is buggy, and nobody wants to touch the JS part.
4) There's no need for the PopoverProvider.
5) We need to stop reinventing the wheel, especially for something as complicated as a popover. Popovers cannot be created solely in C# and mostly rely on JS, which makes it challenging for us to ensure quality.

versile2 commented:
Well butter my biscuit, let me break this down about them popovers objectively!

Here's the question: If we ain't specifically told a popover to flip when it hits the edge, should we be frettin' about it wandering off the screen at all?

Right now, the logic's acting like a cautious cowhand - it won't let popovers get cut off on the top or left unless they're fixing to get completely cut off. But we've been letting them run wild on the right and bottom since:
1. With overflow behavior on, they'll naturally flip like a cattle dog
2. If they're already flipped, there ain't no use trying again since there's no more real estate to work with

Sounds to me like if we ain't specifically told that popover to mind its manners (with overflow behavior), we ought to let it do whatever it pleases - even if that means hanging off the edge of the screen like a wayward tumbleweed.

Would you reckon that's what you were getting at?

Also in regards to 7013 and 7707 they are flipping based on the current logic but still being cut off, perhaps I should run the logic twice to cover both sides if it's out of bounds?
